pkg/pillar/cmd/zedagent/handlecipherconfig.go

```diff
@@ -133,134 +59,196 @@ func parseCipherContextConfig(getconfigCtx *getconfigContext,
 			log.Debugf("parseCipherContextConfig ignoring empty\n")
 			continue
 		}
-		cipherContext := new(types.CipherContext)
-		cipherContext.ID = cfgCipherContext.GetContextId()
-		cipherContext.HashScheme = cfgCipherContext.GetHashScheme()
-		cipherContext.KeyExchangeScheme = cfgCipherContext.GetKeyExchangeScheme()
-		cipherContext.EncryptionScheme = cfgCipherContext.GetEncryptionScheme()
-		cipherContext.DeviceCertHash = cfgCipherContext.GetDeviceCertHash()
-		cipherContext.ControllerCertHash = cfgCipherContext.GetControllerCertHash()
-		log.Debugf("parseCipherContextConfig publishing %v\n", cipherContext)
-		publishCipherContextConfig(getconfigCtx, cipherContext)
+		cipherCtx := types.CipherContext{}
+		cipherCtx.ContextID = cfgCipherContext.GetContextId()
+		cipherCtx.HashScheme = cfgCipherContext.GetHashScheme()
+		cipherCtx.KeyExchangeScheme = cfgCipherContext.GetKeyExchangeScheme()
+		cipherCtx.EncryptionScheme = cfgCipherContext.GetEncryptionScheme()
+		cipherCtx.DeviceCertHash = cfgCipherContext.GetDeviceCertHash()
+		cipherCtx.ControllerCertHash = cfgCipherContext.GetControllerCertHash()
+		updateCipherContextCerts(ctx, &cipherCtx)
+		publishCipherContextConfig(ctx, cipherCtx)
 	}
 }
 
-func publishCipherContextConfig(getconfigCtx *getconfigContext,
-	config *types.CipherContext) {
+// on create/modification, update the cipher context certs
+func updateCipherContextCerts(ctx *getconfigContext,
+	cipherCtx *types.CipherContext) {
+	// get controller cert
+	if len(cipherCtx.ControllerCert) == 0 {
+		cert := getControllerCert(ctx.zedagentCtx,
```

Comment:
Seems like this can return an empty byte array if it isn't found. Don't you need at least an error check, and potentially a way to trigger fetching updated controller certs in that case?

Reply:
Have added the error orchestration through config/status separation for both ControllerCert and CipherContext. Done.